lib/spack/spack/filesystem_view.py

```diff
@@ -434,16 +430,23 @@ def needs_file(spec, file):
                 manifest = {}
             return test_path in manifest
 
-        # remove if dest is not owned by any other package in the view
-        # This will only be false if two packages are merged into a prefix
-        # and have a conflicting file
-
-        # check all specs for whether they own the file. That include the spec
-        # we are currently removing, as we remove files before unlinking the
-        # metadata directory.
-        if len([s for s in self.get_all_specs()
-                if needs_file(s, dest)]) <= 1:
-            os.remove(dest)
+        specs = self.get_all_specs()
+
+        for file in files:
+            if not os.path.lexists(file):
+                tty.warn("Tried to remove %s which does not exist" % file)
+                continue
+                
+            # remove if file is not owned by any other package in the view
+            # This will only be false if two packages are merged into a prefix
+            # and have a conflicting file
+
+            # check all specs for whether they own the file. That include the spec
+            # we are currently removing, as we remove files before unlinking the
+            # metadata directory.
+            if len([s for s in specs if needs_file(s, file)]) <= 1:
```

Comment:
Note this only works because the current code ensures we're only deleting files from a single package at once. If we expand to remove multiple specs at once we will need to revisit this line (and do some sort of reference counting).

Reply:
Glad it looks good. 

We were also worried about that. Currently parallel removes really bork everything, so I figured it wouldn't be an issue to change how it might get borked. I was thinking just a global lock on everything when doing a remove, before doing the tree traversal, would be sufficient to resolve this problem, but that's a discussion for another day.